src/bson_compat.rs

```diff
@@ -1,31 +1,131 @@
-#[cfg(feature = "bson-3")]
-pub(crate) trait RawDocumentBufExt {
-    fn append_ref<'a>(
+use crate::bson::RawBson;
+
+pub(crate) trait RawDocumentBufExt: Sized {
+    fn append_err(&mut self, key: impl AsRef<str>, value: impl Into<RawBson>) -> RawResult<()>;
```

Comment:
Since `append` and similar are now fallible in 3.0, the compatibility layer needed to make them fallible (at least at the type level) for 2.x as well, even if they never actually fail.  The other option would be to wrap the 3.x variants in `expect` but this way people opting in to `bson-3` get nicer error behavior as a bonus :)